File: src/main/java/Enums/Easy.java
package main.java.Enums;

import java.lang.StringBuilder;

public class Easy {
    public enum TrafficSignal {
        RED,
        GREEN,
        YELLOW;

        public String getInstruction() {
            switch (this) {
                case RED:
                    return "STOP";

                case YELLOW:
                    return "SLOW DOWN";

                case GREEN:
                    return "GO";

                default:
                    throw new AssertionError();
            }
        }
    }

    public enum Day {
        MONDAY,
        TUESDAY,
        WEDNESDAY,
        THURSDAY,
        FRIDAY,
        SATURDAY,
        SUNDAY;

        public boolean isWeekend() {
            return this == SATURDAY || this == SUNDAY;
        }
    }

    public enum Direction {
        UP,
        DOWN,
        LEFT,
        RIGHT;

        public static int[] move(int[] start, Direction[] directions) {
            for (Direction direction : directions) {
                switch (direction) {
                    case UP -> start[1]++;
                    case DOWN -> start[1]--;
                    case LEFT -> start[0]--;
                    case RIGHT -> start[0]++;
                }
            } return start;
        }
    }
    
    public enum Chess {
        WHITE,
        BLACK;

        public Chess opposite() {
            return switch (this) {
                case WHITE -> BLACK;
                case BLACK -> WHITE;
            };
        }

        public static String opposite(String color) {
            Chess piece = Chess.valueOf(color);
            return piece.opposite().name();
        }
    }

    public enum Season {
        SPRING,
        SUMMER,
        AUTUMN,
        WINTER;

        public static Season[] getSeasons() {
            return Season.values();
        } 

        public static int countSeasons() {
            return Season.values().length;
        }
    } 
    
    public enum GameMenu {
        NEW_GAME,
        LOAD_GAME,
        SETTINGS,
        EXIT;

        public static String renderMenu() {
            StringBuilder sb = new StringBuilder();
            for (GameMenu gm : GameMenu.values()) {
                String optionName = gm.name().replaceAll("_+", " ");
                sb.append((gm.ordinal()+1) + ". " + optionName + "\n");
            }
                
            return sb.toString();
        }
    }

    public enum GameCommand {
        MOVE,
        ATTACK,
        DEFEND,
        QUIT;

        public static GameCommand parse(String raw) {
            try {
                return GameCommand.valueOf(raw);
            } catch (IllegalArgumentException e) {
                return null;
            }
        }
    }

    public enum Task {
        LOW,
        MEDIUM,
        HIGH,
        CRITICAL;

        private static void swap(Task[] tasks, int x, int y) {
            Task temp = tasks[x];
            tasks[x] = tasks[y];
            tasks[y] = temp;
        }

        public static void sortByPrioritySelectionSort(Task[] tasks) {
            if (tasks.length == 0)
                return;

            Task maxValue;
            int maxIdx;
            for (int i = 1; i < tasks.length; i++) {
                maxIdx = i-1; maxValue = tasks[i-1];
                for (int j = i; j < tasks.length; j++) {
                    if (maxValue.compareTo(tasks[j]) < 0) {
                        maxValue = tasks[j];
                        maxIdx = j;
                    }
                } if (maxIdx != i-1) swap(tasks, maxIdx, i-1);
            }
        }

        public static void sortByPriorityBubbleSort(Task[] tasks) {
            if (tasks.length <= 1)
                return;
            for (int i = tasks.length; i > 1; i--) {
                boolean swapped = false;
                for (int j = 1; j < i; j++) {
                    if (tasks[j].compareTo(tasks[j-1]) > 0) {
                        swap(tasks, j, j-1);
                        swapped = true;
                    }
                }

                if (!swapped)
                    return;
            }
        }

        public static void sortByPriorityInsertionSort(Task[] tasks) {
            if (tasks.length <= 1)
                return;

            for (int i = 1; i < tasks.length; i++) {
                Task temp = tasks[i];

                int j = i-1;
                for (; j >= 0 && tasks[j].compareTo(temp) < 0; j--) {
                    tasks[j+1] = tasks[j];
                }

                tasks[j+1] = temp;
            }
        }

        @Override
        public String toString() {
            return this.name();
        }
    }
}

File: src/main/python/Enums/easy.py
from __future__ import annotations
from enum import Enum, auto

class Easy:
    class TrafficSignal(Enum):
        RED = 1
        YELLOW = 2
        GREEN = 3

        def get_instruction(self) -> str:
            match self:
                case Easy.TrafficSignal.RED:
                    return "STOP"

                case Easy.TrafficSignal.YELLOW:
                    return "SLOW DOWN"

                case Easy.TrafficSignal.GREEN:
                    return "GO"

                case _:
                    raise AssertionError()

    class Day(Enum):
        MONDAY = 1
        TUESDAY = 2
        WEDNESDAY = 3
        THURSDAY = 4
        FRIDAY = 5
        SATURDAY = 6
        SUNDAY = 7

        def is_weekend(self) -> bool:
            return self in (Easy.Day.SATURDAY, Easy.Day.SUNDAY)

    class Direction(Enum):
        UP = (0, 1)
        DOWN = (0, -1)
        LEFT = (-1, 0)
        RIGHT = (1, 0)

        def __init__(self, dx: int, dy: int):
            self.dx = dx
            self.dy = dy

        @staticmethod
        def move(
            start: list[int],
            directions: list[Easy.Direction]
        ) -> tuple[int, int]:

            for direction in directions:
                start[0] += direction.dx
                start[1] += direction.dy

            return tuple(start)

    class Chess(Enum):
        WHITE = auto()
        BLACK = auto()

        def opposite(self) -> Easy.Chess:
            match self:
                case Easy.Chess.WHITE:
                    return Easy.Chess.BLACK

                case Easy.Chess.BLACK:
                    return Easy.Chess.WHITE

                case _:
                    raise AssertionError()

    class Season(Enum):
        SPRING = 1
        SUMMER = 2
        AUTUM = 3
        WINTER = 4

        @staticmethod
        def get_seasons() -> list[Easy.Season]:
            return [season.name for season in Easy.Season];

        @staticmethod
        def count_seasons() -> int:
            return len(Easy.Season);

    class GameMenu(Enum):
        NEW_GAME = 1
        LOAD_GAME = 2
        SETTINGS = 3
        EXIT = 4

        @staticmethod
        def renderMenu() -> str:
            menu = ""
            for option in Easy.GameMenu:
                name = option.name.replace("_", " ")
                menu += f"{option.value}. {name}\n"

            return menu

    class GameCommand(Enum):
        MOVE = auto()
        ATTACK = auto()
        DEFEND = auto()
        QUIT = auto()

        @staticmethod
        def parse(raw:str) -> Easy.GameCommand | None:
            try:
                return Easy.GameCommand[raw]
            except KeyError:
                return None 
        
